File: python/797. All Paths From Source to Target.py
class Solution(object):
    def allPathsSourceTarget(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[List[int]]
        """
        n = len(graph)
        vis = [0]*n
        res = []
        def dfs(cur, path):
            if cur==n-1:
                res.append(path)
                return

            for nx in graph[cur]:
                if vis[nx]: continue
                vis[nx] = 1
                dfs(nx, path+[nx])
                vis[nx] = 0

        dfs(0, [0])
        return res

File: go/src/typical_problems/最短路径/797. All Paths From Source to Target/797. All Paths From Source to Target.go
package main

import "fmt"

/**
https://leetcode-cn.com/problems/all-paths-from-source-to-target/
思路：dfs
 */
func allPathsSourceTarget(graph [][]int) [][]int {
	ans := [][]int{}
	n := len(graph)

	var dfs func(int, []int)
	dfs = func(i int, path []int) {
		if i == n-1 {
			ans = append(ans, append([]int{}, path...))
			return
		}
		for _, ni := range graph[i] {
			path = append(path, ni)
			dfs(ni, path)
			path = path[:len(path) - 1]
		}
	}
	dfs(0, []int{0})
	return ans
}

func main() {
	fmt.Println(allPathsSourceTarget([][]int{{1,2},{3},{3},{}}))
}
